### DISCUSSION TITLE
How to charge the battery during simulation?

### DISCUSSION THREAD

**Discussion by TheASingh** [CURENT/andes#273] [General]
Dr. Cui,

How can I charge the ESD1 battery during simulation?
In below simulation, value of `En` is  0.1. And I ran the simulation for 10s.
At 10s, I updated `Pext0` of ESD1 to -100. But still couldn't charge the battery.

Am I updating the correct variable?
![image](https://user-images.githubusercontent.com/96423656/155162398-70fac597-2e5a-4a67-897f-118da9c180fd.png)

**Comment by cuihantao** [comment]
Did you update it in place using the alter method? 
Double check if the update was successful by checking Psum.

**Reply by TheASingh** [reply]
**1. Below are the values after running the simulation for 10s**

![image](https://user-images.githubusercontent.com/96423656/155181188-e416027f-4643-4695-8bd7-7093b3eee5e6.png)

**2. Below are the values after updating `Pext0 ` to -100 at t= 10s**

![image](https://user-images.githubusercontent.com/96423656/155181678-feee81fc-c617-4b57-bb02-57dbc4c20ed4.png)

**3. And below are the values after continuing the simulation to 20s.**

![image](https://user-images.githubusercontent.com/96423656/155182122-a546fe03-73f3-440c-9777-392744d41e07.png)



**I also get the below error while continuing the simulation to 20s.**

![image](https://user-images.githubusercontent.com/96423656/155192529-91636ee1-a155-46e3-b4aa-be4aca894e51.png)

**Reply by cuihantao** [reply]
Can you post a minimal working example to reproduce the issue? Without the code, it's quite difficult to diagnose. Thanks!

**Reply by cuihantao** [reply]
The value `-1000` seems too large. `Pext0` is per unit in the base of 100 MVA. You can first try to reduce it to, say, `-0.01` and try. There might be other issues, but I cannot tell without looking at the code.

**Reply by TheASingh** [reply]
**-0.01** worked.

![image](https://user-images.githubusercontent.com/96423656/155200039-6e2cd372-6fce-4180-88ef-9fbb9558767a.png)

Could you please tell me one more thing, **why there is no decrease in the `pIG_y ` plot when I kept `En` value to default (100)?**

But when I used 10 instead of 100, `pIG_y ` started decreasing.

**Reply by cuihantao** [reply]
It could be that for En=100, the discharge is so small for 20 seconds, so that SOC change is hardly noticeable.

**Reply by TheASingh** [reply]
Thanks,

Below is my use case:
1. I didn't change in **En** value in ESD1 sheet. And used the default e.g. 100.
2. I changed `SOCinit `value to **0.05** (from 0.5) for **ESD1_1** only.
3. I also updated the concerned **PV** sheet/tab item of **Bus 4** as below:
   `p0  = 0.1`
   `q0  = 0.03`
  ` pmax  = 0.1`
   `pmin  = 0.01`
   `qmax = 0.1`
   `qmin  = -0.06`

Query: Please guide me how to discharge this **ESD1_1** item if above updated values are correct?
Note: I ran the TDS for 2000s for above use case, but there was no change in `pIG_y` plot.

**Reply by cuihantao** [reply]
Please post code and case files that I can replicate the issue.

**Reply by TheASingh** [reply]
[ieee14_dgprct1_solar_ESD1.xlsx](https://github.com/cuihantao/andes/files/8126962/ieee14_dgprct1_solar_ESD1.xlsx)

I have attached the above input file. And I haven't made any changes in the code files.

I tried one more time with tstep = 0.33, and I was able to see the change in `pIG_y`.
But when I tried with default tstep, I'm still facing the issue.

**Below result is for tstep = 0.33**

![image](https://user-images.githubusercontent.com/96423656/155386894-efed3bdb-cd02-4181-8f17-3a329f753231.png)

**Below result is for default tstep e.g. 0.033** 

![image](https://user-images.githubusercontent.com/96423656/155386948-3fbddb0a-5971-4e6d-82b2-4b295cba2a93.png)

**Reply by cuihantao** [reply]
Sorry but I'm not sure; where to find the code?

**Reply by cuihantao** [reply]
Do you mean the example notebook file?

**Reply by cuihantao** [reply]
It would be really helpful if you can send a block of code that can run and replicate the issue. Thanks.

**Comment by TheASingh** [comment]
[ieee14_dgprct1_solar_ESD1.xlsx](https://github.com/cuihantao/andes/files/8127336/ieee14_dgprct1_solar_ESD1.xlsx)
Please run the following commands one by one on **Jupyter**. And please use the above excel file as input.
1. **Code for tstep = 0.33**

```
import andes
from andes.utils.paths import get_case
andes.config_logger(stream_level=30)
```

_Below you need to provide the path of the input file (excel sheet that I have provided)_
```
ss= andes.load('C:/Users/ASingh/.conda/pkgs/andes-1.5.8-pyhd8ed1ab_0/site-packages/andes/cases/ieee14/ieee14_dgprct1_solar_ESD1.xlsx', setup=False, no_output=True)
ss.setup()
ss.PFlow.run()
ss.PVD1.Iqcmd.lim.allow_adjust

ss.TDS.config.tf = 1800
ss.TDS.config.tstep = 0.33
ss.TDS.config.shrinkt = 0
ss.TDS.run()
ss.TDS.plt.plot(ss.ESD1.pIG_y, a=(0), ylabel='Test', greyscale=True, grid=True, savefig=False)
```





2. **Code for default tstep**

```
import andes
from andes.utils.paths import get_case
andes.config_logger(stream_level=30)
```

_Below you need to provide the path of the input file (excel sheet that I have provided)_
```
ss= andes.load('C:/Users/ASingh/.conda/pkgs/andes-1.5.8-pyhd8ed1ab_0/site-packages/andes/cases/ieee14/ieee14_dgprct1_solar_ESD1.xlsx', setup=False, no_output=True)
ss.setup()
ss.PFlow.run()
ss.PVD1.Iqcmd.lim.allow_adjust

ss.TDS.config.tf = 1800
ss.TDS.run()
ss.TDS.plt.plot(ss.ESD1.pIG_y, a=(0), ylabel='Test', greyscale=True, grid=True, savefig=False)
```


Let me know, if you still not get anything.

**Reply by cuihantao** [reply]
Thanks! I will get back to you soon.

**Comment by cuihantao** [comment]
It appears to be a bug! The change in SOC is too small and swallowed by numerical errors. Let me fix it.

**Comment by cuihantao** [comment]
This issue is fixed in my `develop` branch. You can also use very large step sizes like `tstep=2`. Previously, it will force it back to 0.033.

A new version will be tagged soon. Thanks!

**Reply by TheASingh** [reply]
Thank you!

Please let me know when the new version is released and is ready to use.

**Reply by cuihantao** [reply]
The new version, v1.5.11 is out with this issue fixed. You can upgrade your package with `conda update andes`. Double check after the upgrade for the version number.

**Reply by TheASingh** [reply]
Thank you very much, the issue is resolved now.

**Comment by TheASingh** [comment]
I have another query.
**What are the min and max values for `pref0` to charge the battery?**

The battery has following parameters:
`En`= 100
`SOCinit`= 0.05,
`p0`(of concerned PV tab item) = 0.1 (As I have 10 batteries, so for one battery, it will be 0.01)
pmax (of concerned PV tab item) = 0.1
pmin (of concerned PV tab item) = -0.5

1. I updated `pref0` to **-0.9**  at 900s. (Straight line)
2. Updated `pref0` to **-0.01**  at 1800s. (Started charging)
3. Updated `pref0` to **-0.02**  at 2700s. (Straight line)
4. Updated `pref0` to **-0.005**  at 3000s. (Started charging)

**What are the min and max values for `pref0` to charge the battery? And also is it correct way to charge the battery?**

![image](https://user-images.githubusercontent.com/96423656/155570663-bbf4956a-ebea-4f42-8471-4cfafe205bbb.png)

**Comment by TheASingh** [comment]
I further verified the impact of `Psum` and results are as follow:
SOCinit = 0.5, En = 100, p0 = 0.1 (combined of 10 ESD1s in PV sheet)

1. Updated `Pext0` to **-0.02** at 300s and ran the TDS (300s to 600s). `Psum` was **-0.00939823**  (Charging)
2. Updated `pref0` to **-0.01** at 600s and ran the TDS (600s to 900s). `Psum` was **-0.0296964**     (Straight Line)
3. Updated `Pext0` to **-0.01** at 900s and ran the TDS (900s to 1200s). `Psum` was **-0.0196964**   (Straight Line)
4. Updated `Pext0` to **0** at 1200s and ran the TDS (1200s to 1500s). `Psum` was **-0.00939823**    (Charging)
5. Updated `Pext0` to **-0.02** and `pref0` to **0.01** at 1500s and ran the TDS (1500s to 1800s). `Psum` was **-0.00939823**  (Charging)
6. Updated `Pext0` to **-0.01** at 1800s and ran the TDS (1800s to 2100s). `Psum` was **0.00029466**  (Discharging)

![image](https://user-images.githubusercontent.com/96423656/155766829-9275ae00-6f8e-470a-9203-7c8f6bee5173.png)

My query is: 
**Is it the Psum only that determines Charging/Discharging/Straight Line? If yes, Please let me know what is the range of Psum for Charging, Discharging and Straight Line?**

**Reply by jinningwang** [reply]
`Psum` should be the most related on that determine the Charging/Discharging of ESD1.

The implementation of ESD1 is basically adding an SoC module to the PVD1. As for PVD1, you can refer to its [doc](https://www.powerworld.com/WebHelp/Content/TransientModels_HTML/Machine%20Model%20PVD1.htm) from PowerWorld.

Another possible factor is the protection function integrated into ESD1. It is controlled by `recflag`, to enable voltage and frequency recovery limiters. The picture is PVD1 control diagram, but ESD1 is similar to it. The module in red rectangles are the voltage and frequency recovery limiters.
<img width="931" alt="image" src="https://user-images.githubusercontent.com/79226045/155802510-ef326c5f-1f52-4f18-957a-460e62aebc80.png">
